com.unity.multiplayer.mlapi/Runtime/Core/NetworkUpdateLoop.cs

```diff
@@ -0,0 +1,525 @@
+        {
+            public static PlayerLoopSystem CreateLoopSystem()
+            {
+                return new PlayerLoopSystem
+                {
+                    type = typeof(NetworkFixedUpdate),
+                    updateDelegate = RunNetworkFixedUpdate
+                };
+            }
+        }
+
+        private struct NetworkPreUpdate
+        {
+            public static PlayerLoopSystem CreateLoopSystem()
+            {
+                return new PlayerLoopSystem
+                {
+                    type = typeof(NetworkPreUpdate),
+                    updateDelegate = RunNetworkPreUpdate
+                };
+            }
+        }
+
+        private struct NetworkUpdate
+        {
+            public static PlayerLoopSystem CreateLoopSystem()
+            {
+                return new PlayerLoopSystem
+                {
+                    type = typeof(NetworkUpdate),
+                    updateDelegate = RunNetworkUpdate
+                };
+            }
+        }
+
+        private struct NetworkPreLateUpdate
+        {
+            public static PlayerLoopSystem CreateLoopSystem()
+            {
+                return new PlayerLoopSystem
+                {
+                    type = typeof(NetworkPreLateUpdate),
+                    updateDelegate = RunNetworkPreLateUpdate
+                };
+            }
+        }
+
+        private struct NetworkPostLateUpdate
+        {
+            public static PlayerLoopSystem CreateLoopSystem()
+            {
+                return new PlayerLoopSystem
+                {
+                    type = typeof(NetworkPostLateUpdate),
+                    updateDelegate = RunNetworkPostLateUpdate
+                };
+            }
+        }
+
+        /// <summary>
+        /// Registers a network update system to be executed in all network update stages.
+        /// </summary>
+        public static void RegisterAllNetworkUpdates(this INetworkUpdateSystem updateSystem)
+        {
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.Initialization);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.EarlyUpdate);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.FixedUpdate);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.PreUpdate);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.Update);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.PreLateUpdate);
+            RegisterNetworkUpdate(updateSystem, NetworkUpdateStage.PostLateUpdate);
+        }
+
+        /// <summary>
+        /// Registers a network update system to be executed in a specific network update stage.
+        /// </summary>
+        public static void RegisterNetworkUpdate(this INetworkUpdateSystem updateSystem, NetworkUpdateStage updateStage = NetworkUpdateStage.Update)
+        {
+            switch (updateStage)
+            {
+                case NetworkUpdateStage.Initialization:
```

Comment:
All those multiple case do the same thing, just on different list. If we were to change the place where update could happen it would be way too much code to adjust.

Instead of having  m_Initialization_List, m_EarlyUpdate_List as separate variables, can we have an associative container, or just an array, with the key being NetworkUpdateStage ?

`m_updateLists List<INetworkUpdateSystem>[]` indexed by `NetworkUpdateStage` ?

The idea being that the update stages should be modifiable without duplicating further any block of code.

Reply:
Good question Jeffrey, I think the idea was to just have a pre-determined set of update stages that provided plenty of options for the user to select from.  However, you do make a good point that perhaps to further improve upon this we might think about a way to initialize the network update loop's defined update stages prior to any INetworkUpdateSystem implemented class registers.